Improve portability

smortex opened the pull request:
The control-repo code is not usable out-of-the-box on FreeBSD (and probably any system where the AIO package is not used).

* Since `bash` is not a requirement (the shell script is a portable bourne shell script), switch to plain `sh` which is always installed in `/bin/`.
* Instead of guessing the path to `ruby`, check if a version installed by the AIO package is available. If so, use this version, otherwise, use `$PATH` to locate Ruby and run the appropriate script.

smortex committed:
Improve portability

- Fix shebang: `bash` is not always in `/bin/`, and since the script
  does not have bashism, rely on `sh` which is always in `/bin/`;
- Use `/opt/puppetlabs/puppet/bin/ruby` if this file exist and is
  executable, otherwise use `ruby` from $PATH;
- Use `code_manager_config_version.rb` if `.r10k-deploy.json` is found,
  and `config_version.rb` in all other cases.

smortex committed:
Remove execute permission and shebangs

The appropriate ruby interpreter is determined by the config_version.sh
shell script which explicitely use it to start these ruby scripts.
Removing the execute bit ensure users will not run these script with the
wrong Ruby version.

smortex committed:
Do sanity checking early

Fail before doing anything when passed parameters are invalid.

smortex committed:
Remove code dead for more than 2 years

r10k generates a .r10k-deploy.json file since version 2.1.0 which was
released on October 28, 2015.  New users of the control-repo are not
likely to have a so old version of r10k, so remove this dead code.

smortex commented:
Hum, r10k ≥ 2.1.0 creates a `.r10k-deploy.json` file, so you should run a version of r10k that is more than 2 years old when you create your control-repo for the `.r10k-deploy.json`-less case to make sense…

Let's cleanup this too 😉

smortex commented:
Rebased on top of production to fix conflict with 36a278427582e6629141eb8539f658cf1604a528.

npwalker commented:
@smortex I don't see a large advantage to removing compatibility with old r10k versions or using straight up git.  

If the PR is stripped down to just making it work on Free BSD I think we'll get more traction on it.

mtancoigne commented:
Hello !

Any news on this ? I'd really like to see FreeBSD support !

joshcooper commented:
I can't seem to ask @puppetlabs/cs-arch for a review, so pinged some individual folks

reidmv reviewed (dismissed):
Looks like this PR does clean up old codepaths that wouldn't have been followed for a long time now. Doing so makes other possible optimizations more visible perhaps, but IMO there's no reason we shouldn't just merge this cleanup now.

reidmv committed:
Restore and clarify config_version functionality

This branch is intended as a portability fix. Some functionailty had
been inadvertently removed as unused, but testing revealed that it had a
purpose. Because the purpose was unclear, this commit restores the
functionality AND clarifies it in the script names and comments in
config_version.sh.

reidmv reviewed (approved):
Added some modifications. The .r10k-deploy.json file thing seems a little more complicated, so to keep the focus on portability for this PR I've put it back. It was confusing what the check was doing though so I've also added some comments and file renames to clarify what's going on.

smortex commented on scripts/config_version.sh:
```diff
@@ -1,14 +1,35 @@
+
+elif [ -x /usr/bin/git ]; then
+  # The git command is available.
+  /usr/bin/git --git-dir $1/$2/.git rev-parse HEAD
```
`git(1)` might be installed in another directory (e.g. `/usr/local/bin/git` on FreeBSD).  Maybe it would be better to assume the directory is in `$PATH` instead of using the full executable name?

reidmv replied:
Good call. Updating to use `type git` to check for the command, and then expecting it on PATH.

smortex replied:
I was about to suggest another minor change and recall it's a branch in my repo. I ammended your commit to avoid relying on `test(1)` (`[(1)`) since `type` exit code tells us if the command was found :wink:

reidmv replied:
Yep, you're totally right. Heh; oops.

reidmv committed:
Don't assume git is in /usr/bin; expect it on PATH
